### PR TITLE
Fix deleting the Broker ConfigMap prevents the Broker from being finalized

### LINKED ISSUE #1580: Deleting the Broker ConfigMap prevents the Broker from being finalized
**Describe the bug**

Deleting the Broker ConfigMap prevents the Broker from being finalized since we can't get
Kafka cluster URL and credentials to remove topics.

 **Expected behavior**

A Broker can be deleted even if the ConfigMap is deleted first.

 **To Reproduce**

- Create a Broker `b`
- Remove the ConfigMap associated with `broker.spec.config`
- Remove the Broker `b`

 **Knative release version**

`<= devel`

 **Additional context**

/area broker
/area control-plane
/kind bug
/priority important-soon

### PR DESCRIPTION
**Pull request by pierDipi**
Deleting the Broker ConfigMap prevents the Broker from being finalized
since we can't get Kafka cluster URL and credentials to remove topics.

In this patch, we're adding a finalizer to the ConfigMap associated
with the Broker and then remove it once the broker is properly
finalized.

Fixes #1580 

## Proposed Changes

- Fix deleting the Broker ConfigMap prevents the Broker from being finalized

**Release Note**

```release-note
Fix deleting the Broker ConfigMap prevents the Broker from being finalized
```

**Docs**

### CODE REVIEW

**Comment by matzew** [comment]
The `has...`, `remove...` and `add...` stand out - perhaps something more like:
https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/controller/controllerutil/controllerutil.go#L351-L375 ?

**Comment by devguyio** [comment]
LGTM
just one NIT: I wonder if it's a good idea to add some debug logging that reflect the lifecyle of the finalizers (added, removed, found etc) cause that'd be a great help debugging nasty finalizers problems.

**Comment by devguyio** [comment]
> The `has...`, `remove...` and `add...` stand out - perhaps something more like: https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/controller/controllerutil/controllerutil.go#L351-L375 ?

Care to elaborate more what do you mean by stad out? not sure I understand what you mean

**Review by devguyio — APPROVED** [review]

File: control-plane/pkg/reconciler/broker/broker.go [hunk 1]
```diff
@@ -391,3 +401,49 @@ func (r *Reconciler) reconcilerBrokerResource(ctx context.Context, topic string,
...
+		}
+	}
+	return nil
+}
+
+func containsFinalizerCM(cm *corev1.ConfigMap, finalizer string) bool {
+	if cm == nil {
+		return false
+	}
+	for _, f := range cm.Finalizers {
```
Comment (matzew) [inline control-plane/pkg/reconciler/broker/broker.go:421]: should we use accessors, like `GetFinalizers()` ?

**Comment by matzew** [comment]
> @matzew Care to elaborate more what do you mean by stad out? not sure I understand what you mean

yeah - part of me wonders if this could be added else for a more generic approach... (I have always that feeling when we add _private_ utils to a file, that is usually supposed to do different).

I've checked some "related" code, but all kinda same (or well, different).  Let's ship this

File: control-plane/pkg/reconciler/broker/broker.go [hunk 2]
```diff
@@ -391,3 +401,49 @@ func (r *Reconciler) reconcilerBrokerResource(ctx context.Context, topic string,
...
 	return resource, nil
 }
+
+func (r *Reconciler) addFinalizerCM(ctx context.Context, finalizer string, cm *corev1.ConfigMap) error {
+	if !containsFinalizerCM(cm, finalizer) {
+		cm := cm.DeepCopy() // Do not modify informer copy.
+		cm.Finalizers = append(cm.Finalizers, finalizer)
+		_, err := r.KubeClient.CoreV1().ConfigMaps(cm.GetNamespace()).Update(ctx, cm, metav1.UpdateOptions{})
+		if err != nil && !apierrors.IsNotFound(err) {
+			return fmt.Errorf("failed to add finalizer to ConfigMap %s/%s: %w", cm.GetNamespace(), cm.GetName(), err)
```
Comment (matzew) [inline control-plane/pkg/reconciler/broker/broker.go:411]: this could be aligned w/ the `remove`

**Comment by matzew** [comment]
None of my comments are blockers - I just hate have provblems w/ adding CRUD code :smile:

**Comment by devguyio** [comment]
> > @matzew Care to elaborate more what do you mean by stad out? not sure I understand what you mean
> 
> yeah - part of me wonders if this could be added else for a more generic approach... (I have always that feeling when we add _private_ utils to a file, that is usually supposed to do different).
> 
> I've checked some "related" code, but all kinda same (or well, different). Let's ship this

gotcha 👍🏽 thanks for elaborating

**Pull request merged**